### PR TITLE
[lldb] Detect a Darwin kernel issue and work around it

### PR DESCRIPTION
**Pull request by jasonmolenda**
On arm64 machines, when there is a hardware breakpoint or watchpoint set, and lldb has instruction-stepped a thread, and then done a Process::Resume, we will sometimes receive an extra "instruction step completed" mach exception and the pc has not advanced.  From a user's perspective, they hit Continue and lldb stops again at the same spot.  From the testsuite's perspective, this has been a constant source of testsuite failures for any test using hardware watchpoints and breakpoints, the arm64 CI bots seem especially good at hitting this issue.

Jim and I have been slowly looking at this for a few months now, and finally I decided to try to detect this situation in lldb and silently resume the process again when it happens.

We were already detecting this "got an insn-step finished mach exception but this thread was not instruction stepping" combination in StopInfoMachException where we take the mach exception and create a StopInfo object for it.  We had a lot of logging we used to understand the failure as it was hit on the bots in assert builds.

This patch adds a new case to `Thread::GetPrivateStopInfo()` to call the StopInfo's (new) `IsContinueInterrupted()` method.  In StopInfoMachException, where we previously had logging for assert builds, I now note it in an ivar, and when `Thread::GetPrivateStopInfo()` asks if this has happened, we check all of the combination of events that this comes up:  We have a hardware breakpoint or watchpoint, we were not instruction stepping this thread but got an insn-step mach exception, the pc is the same as the previous stop's pc.  And in that case, `Thread::GetPrivateStopInfo()` returns no StopInfo -- indicating that this thread would like to resume execution.

The `Thread` object has two StackFrameLists, `m_curr_frames_sp` and `m_prev_frames_sp`.  When a thread resumes execution, we move `m_curr_frames_sp` in to `m_prev_frames_sp` and when it stops executing, w euse `m_prev_frames_sp` to seed the new `m_curr_frames_sp` if most of the stack is the same as before.

In this same location, I now save the Thread's RegisterContext::GetPC into an ivar, `m_prev_framezero_pc`.  StopInfoMachException needs this information to check all of the conditions I outlined above for `IsContinueInterrupted`.

This has passed exhaustive testing and we do not have any testsuite failures for hardware watchpoints and breakpoints due to this kernel bug with the patch in place.  In focusing on these tests for thousands of runs, I have found two other uncommon race conditions for the TestConcurrent* tests on arm64.  TestConcurrentManyBreakpoints.py (which uses no hardware watchpoint/breakpoints) will sometimes only have 99 breakpoints when it expects 100, and any of the concurrent tests using the shared harness (I've seen it in
TestConcurrentWatchBreakDelay.py, TestConcurrentTwoBreakpointsOneSignal.py, TestConcurrentSignalDelayWatch.py) can fail when the test harness checks that there is only one thread still running at the end, and it finds two -- one of them under pthread_exit / pthread_terminate. Both of these failures happen on github main without my changes, and with my changes - they are unrelated race conditions in these tests, and I'm sure I'll be looking into them at some point if they hit the CI bots with frequency.  On my computer, these are in the 0.3-0.5% of the time class.  But the CI bots do have different timing.

### CODE REVIEW

File: lldb/include/lldb/Target/StopInfo.h [hunk 1]
```diff
@@ -79,6 +79,11 @@ class StopInfo : public std::enable_shared_from_this<StopInfo> {
 
   virtual bool IsValidForOperatingSystemThread(Thread &thread) { return true; }
 
+  /// A Continue operation can result in a false stop event
+  /// before any execution has happened in certain cases on Darwin.
+  /// We need to silently continue again time.
```
Comment (bulbazord) [inline lldb/include/lldb/Target/StopInfo.h:84]: "continue again time." -> "continue again one more time."

Is this what you meant?

File: lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp [hunk 2]
```diff
@@ -825,6 +806,56 @@ StopInfoSP StopInfoMachException::CreateStopReasonWithMachException(
     break;
   }
 
-  return StopInfoSP(new StopInfoMachException(thread, exc_type, exc_data_count,
-                                              exc_code, exc_sub_code));
+  return StopInfoSP(new StopInfoMachException(
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp:809]: Since you're already here: `make_shared`? 😄

**Review by bulbazord — COMMENTED** [review]
Makes sense to me!

File: lldb/include/lldb/Target/StopInfo.h (hunk 1 above)
Comment (jimingham) [inline lldb/include/lldb/Target/StopInfo.h:84]: "again time" -> "again one more time"?

Also, this isn't a Darwin specific feature, on any system where the Continue might stall, you could use this.  I'd describe what it does rather than primarily refer to Darwin.

File: lldb/include/lldb/Target/StopInfo.h [hunk 3]
```diff
@@ -79,6 +79,11 @@ class StopInfo : public std::enable_shared_from_this<StopInfo> {
...
+  virtual bool IsContinueInterrupted(Thread &thread) { return false; }
```
Comment (jimingham) [inline lldb/include/lldb/Target/StopInfo.h:85]: This may be a little picky, but the continue that was interrupted is in the past by the time you call this.  WasContinueInterrupted reads more naturally to me because of that.

File: lldb/include/lldb/Target/Thread.h [hunk 4]
```diff
@@ -1226,6 +1227,7 @@ class Thread : public std::enable_shared_from_this<Thread>,
   friend class StackFrameList;
   friend class StackFrame;
   friend class OperatingSystem;
+  friend class StopInfoMachException;
```
Comment (jimingham) [inline lldb/include/lldb/Target/Thread.h:1230]: Why is this necessary?  GetPreviousFrameZeroPC is public, right?

File: lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp [hunk 5]
```diff
@@ -825,6 +806,56 @@ StopInfoSP StopInfoMachException::CreateStopReasonWithMachException(
...
+      thread, exc_type, exc_data_count, exc_code, exc_sub_code,
+      not_stepping_but_got_singlestep_exception));
+}
+
+// Detect an unusual situation on Darwin where:
+//
+//   0. We did an instruction-step before this.
+//   1. We have a hardware breakpoint or watchpoint set.
+//   2. We are resuming the process.
```
Comment (jimingham) [inline lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp:818]: Technically we also "resume" the process when we single step.  You mean we're not single stepping here, I'd say the explicitly.

**Review by jimingham — CHANGES_REQUESTED** [review]
This looks good.  You're being careful to do as narrow as possible a check for the buggy behavior.  Handling this in the private StopInfo is great, that renders it a non-event in the higher level ShouldStop negotiations, which is exactly what we want.

I had a couple trivial comments, and I don't think you should need the `friend class` statement.  If you do need that for reasons I missed, we should probably make some API to get around that, StopInfoMachException really shouldn't need private access to the Thread class.

**Comment by jasonmolenda** [comment]
Thanks for all the feedback @bulbazord @jimingham , updated the PR with those issues addressed.

File: lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp [hunk 6]
```diff
@@ -825,6 +806,56 @@ StopInfoSP StopInfoMachException::CreateStopReasonWithMachException(
...
+
+// Detect an unusual situation on Darwin where:
+//
+//   0. We did an instruction-step before this.
+//   1. We have a hardware breakpoint or watchpoint set.
+//   2. We resumed the process, but not with an instruction-step.
+//   3. The thread gets an "instruction-step completed" mach exception.
+//   4. The pc has not advanced - it is the same as before.
+//
+// This method returns true for that combination of events.
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp:822]: It might be nice to make these /// and move them into the header file. This way doxygen can pick it up.

File: lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp [hunk 7]
```diff
@@ -825,6 +806,56 @@ StopInfoSP StopInfoMachException::CreateStopReasonWithMachException(
...
+    return false;
+
+  // The previous pc value and current pc value are the same.
+  if (*prev_pc != reg_ctx_sp->GetPC())
+    return false;
+
+  // We have a watchpoint -- this is the kernel bug.
+  ProcessSP process_sp = thread.GetProcess();
+  if (process_sp->GetWatchpointResourceList().GetSize()) {
+    LLDB_LOGF(log,
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp:843]: This can be the more modern LLDB_LOG

**Review by jimingham — APPROVED** [review]
Thanks for taking out the friend.  LGTM.
